daemon/containerd/image_pull.go

```diff
@@ -40,6 +40,27 @@ func (i *ImageService) PullImage(ctx context.Context, baseRef reference.Named, p
 	}()
 	out := streamformatter.NewJSONProgressOutput(outStream, false)
 
+	// How long downloaded content will be held after cancellation
+	expireAt := time.Now().Add(8 * time.Hour)
+	ctx, done, err := i.client.WithLease(ctx,
+		leases.WithRandomID(),
+		leases.WithLabels(map[string]string{
+			"moby/action.pull":        baseRef.String(),
+			"containerd.io/gc.expire": expireAt.Format(time.RFC3339),
```

Comment:
Hm... I was surprised that these were not exported as either a `const` or as a utility in containerd; should they be? https://github.com/moby/moby/blob/803534fb4e784d3e6bef9846d64ce4960ad1ec80/vendor/github.com/containerd/containerd/metadata/gc.go#L60-L65

I see containerd v2 added some extra documentation to some of these to outline the required format for values set; https://github.com/containerd/containerd/blob/f0ebbd3c8a662656b458e7e1cca059a690d557a7/core/metadata/gc.go#L69-L76

```go
	// labelGCExpire indicates that an object is collectible after the
	// provided time. For image objects, this makes them available to
	// garbage collect when expired, when not provided, image objects
	// are root objects that never expire. For non-root objects such
	// as content or snapshots, these objects will be treated like
	// root objects before their expiration.
	// Expected format is RFC 3339
	labelGCExpire = []byte("containerd.io/gc.expire")
```


So maybe a utility could make most sense;

- `leases.WithGCExpireAt(time.Time)` (or something) to make sure it's in the correct format
- Some of the other ones may require a specific format as well, so _perhaps_ could require a utility as well to validate (instead of a plain `const` for the right label)?

Reply:
There is a utility, here it is just to be specific. There is no logic around it though and the format doesn't change, seemed cleaner to just clearly see the labels that are getting set.

Reply:
Oh, I see there's indeed a `WithExpiration`; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/vendor/github.com/containerd/containerd/leases/lease.go#L81-L91


Yeah, mostly looking that using a consistent approach may help finding related things. In this case, I just checked where we're setting these, and I found these;

```bash
git grep '\.WithExpiration' -- ':!vendor'
builder/builder-next/adapters/containerimage/pull.go:   ctx, done, err := leaseutil.WithLease(ctx, p.is.LeaseManager, leases.WithExpiration(5*time.Minute), leaseutil.MakeTemporary)
daemon/containerd/image_builder.go:     ctx, _, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/containerd/image_builder.go:             leases.WithExpiration(time.Hour*24),
daemon/containerd/image_builder.go:     ctx, release, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/containerd/image_commit.go:      ctx, release, err := i.client.WithLease(ctx, leases.WithRandomID(), leases.WithExpiration(1*time.Hour))
daemon/images/image_pull.go:            leases.WithExpiration(24 * time.Hour),
```

Interestingly it looks like for non-containerd image-pulls, we still use 24 hours, but a different label; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/daemon/images/image_pull.go#L118-L125

And for classic builder with containerd as well, and again with a different label; https://github.com/moby/moby/blob/113d3cecea2c0a6181c34489888a8a89de4e1961/daemon/containerd/image_builder.go#L235-L241


Perhaps those should be corrected as well?

Reply:
I'm not going to touch the ones outside of daemon/containerd for now, those might be worth looking but not related to the issues we are seeing with the containerd integration. I added a `withLease` method to our own image service that will add the labels and logic that is consistent with pruning.